File: src/engine.ts
import { LetterResult } from "./interface";
import { getWordWeightage } from "./utils";
import listOfWords from "./wordList";

class WordGuessEngine {
  filteredWordList: string[];
  letterInformation: Record<string, LetterResult>;

  constructor() {
    this.filteredWordList = listOfWords;
    this.letterInformation = {};
  }

  // uses the reverse engineered assumption as a checker
  static getCorrectGuessOfDay(): string {
    const startDay = new Date(2021, 5, 19, 0, 0, 0, 0);
    const today = new Date();
    const index = Math.round(
      (today.setHours(0, 0, 0, 0) - startDay.setHours(0, 0, 0, 0)) / 864e5
    );
    return listOfWords[index % listOfWords.length];
  }

  getGuess(): string {
    let bestGuess = "";
    let bestGuessWeight = 0;
    this.filteredWordList.forEach((word) => {
      const wordWeight = this.wordWeightage(word);
      if (wordWeight > bestGuessWeight) {
        bestGuessWeight = wordWeight;
        bestGuess = word;
      }
    });
    console.log(`Guessing ${bestGuess} with a weightage of ${bestGuessWeight}`);
    return bestGuess;
  }

  //TODO: Implement weightage based on information in this.letterInformation
  wordWeightage(word: string): number {
    return getWordWeightage(word);
  }
}

export default WordGuessEngine;

File: src/python/engine.py
import datetime
from worldList import LIST_OF_ANSWERS
from interface import LetterResult
from utils import *

class Engine:
    def __init__(self) -> None:
        self.filteredWordList = LIST_OF_ANSWERS
    
    def getCorrectGuessOfDay() -> str:
        start_day = datetime.date(2021, 6, 19)
        today = datetime.date.today()
        index = (today - start_day).days
        return LIST_OF_ANSWERS[index % len(LIST_OF_ANSWERS)]

    def getGuess(self) -> str :
        bestGuess: str = ''
        bestGuessWeight: int = 0
        for word in self.filteredWordList:
            wordWeight = getWordWeightage(word)
            if(wordWeight > bestGuessWeight):
                bestGuess = word
                bestGuessWeight = wordWeight
        return bestGuess

    def handleGuessResult(self, word: str, result:str) -> None:
        guessResult = parseGuessResult(word, result)

        for letter in guessResult.keys():
            res : LetterResult = guessResult[letter]

            if res.status == "INCORRECT":
                self.filteredWordList = list(
                    filter(lambda word: not letter in word, self.filteredWordList)
                )
            else:
                self.filteredWordList = list(
                    filter(
                        lambda word: filterWord(letter, word, res.correctPositions, res.incorrectPositions),
                        self.filteredWordList 
                    )
                )
        return
